mozilla_version/mobile.py

```diff
@@ -42,30 +46,97 @@ class MobileVersion(BaseVersion):
     _VALID_ENOUGH_VERSION_PATTERN = re.compile(r"""
         ^(?P<major_number>\d+)
         \.(?P<minor_number>\d+)
-        \.(?P<patch_number>\d+)
+        (\.(?P<patch_number>\d+))?
         (
-            -beta\.(?P<beta_number>\d+)
+            (?P<is_nightly>a1)
+            |(-beta\.|b)(?P<beta_number>\d+)
             |-rc\.(?P<release_candidate_number>\d+)
-        )?$""", re.VERBOSE)
+        )?
+        -?(build(?P<build_number>\d+))?$""", re.VERBOSE)
 
-    # Patch_number is required, so can't be none like in BaseVersion
-    patch_number = attr.ib(type=int, converter=positive_int, default=None)
-
-    _MANDATORY_NUMBERS = BaseVersion._MANDATORY_NUMBERS + (
-        'patch_number',
-    )
     _OPTIONAL_NUMBERS = (
-        'beta_number', 'release_candidate_number',
+        'patch_number', 'beta_number', 'release_candidate_number', 'build_number'
     )
 
-    _ALL_NUMBERS = _MANDATORY_NUMBERS + _OPTIONAL_NUMBERS
+    _ALL_NUMBERS = BaseVersion._MANDATORY_NUMBERS + _OPTIONAL_NUMBERS
+
+    _FIRST_VERSION_TO_FOLLOW_GECKO_PATTERN = 104
 
+    build_number = attr.ib(type=int, converter=strictly_positive_int_or_none, default=None)
     beta_number = attr.ib(type=int, converter=strictly_positive_int_or_none, default=None)
+    is_nightly = attr.ib(type=bool, default=False)
     release_candidate_number = attr.ib(
         type=int, converter=strictly_positive_int_or_none, default=None
     )
     version_type = attr.ib(init=False, default=attr.Factory(_find_type, takes_self=True))
 
+    def __attrs_post_init__(self):
+        """Ensure attributes are sane all together."""
+        error_messages = []
+
+        if self.is_gecko_pattern:
+            error_messages.extend([
+                pattern_message
+                for condition, pattern_message in ((
+                    self.beta_number is not None and self.patch_number is not None,
+                    'Beta number and patch number cannot be both defined',
+                ), (
+                    self.release_candidate_number is not None,
+                    'Release candidate number cannot be defined starting Mobile v104',
```

Comment:
nit: Error message seems a little awkward. Maybe "...starting with Mobile v104" or "...after..."?

Reply:
Thank you very much! I do enjoy this kind of feedback. It helps me speak/write better English every day 👍 Last year, I actually read this eBook[1] and I was like "ohhhh, I've made this mistake for so many years!". 

So, keep this feedback coming! 😃 

[1] https://jakubmarian.com/english-mistakes/